openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go

```diff
@@ -105,173 +109,197 @@ func (o *DiscoverEtcdInitialClusterOptions) Validate() error {
+		// Result: provide initial cluster value so etcd can communicate with peers during startup.
+		if found && targetMember.Name == "" && !dataDirExists {
+			fmt.Print(getInitialCluster(o.TargetName, targetMember, cluster.Members))
+			return nil
 		}
-		fmt.Fprintf(os.Stderr, "      target=%v, err=%v\n", stringifyMember(targetMember), err)
 
-		// we're done because we found what we want.
-		if targetMember != nil && err == nil {
-			break
+		// Condition: unstarted member found with dataDir
+		// This member is part of the cluster but has not yet started yet has a dataDir
+		// Result: archive old dataDir and return Error
+		if found && targetMember.Name == "" && dataDirExists {
+			archivedDir, err := archiveDataDir(o.DataDir)
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("member %q is unstarted but previous members dataDir exists: archiving to %q", target.String(), archivedDir)
 		}
 
-		fmt.Fprintf(os.Stderr, "#### sleeping...\n")
-		time.Sleep(1 * time.Second)
-	}
+		// Condition: member found and has started in the past
+		// Result: empty initial config
+		if found && dataDirExists {
+			return nil
+		}
 
-	switch {
-	case targetMember == nil && memberDirExists:
-		// we weren't able to locate other members and need to return based previous memberDir so we can restart.  This is again the off and on flow.
-		fmt.Fprintf(os.Stderr, "Couldn't get targetMember, but memberDir %s is present on %s. Returning with no error.\n", memberDir, o.TargetName)
-		fmt.Printf(o.TargetName)
-		return nil
+		// Condition: member found and has started in the past but not has no dataDir
+		// A member is not actually gone forever unless it is removed from MemberList or its dataDir is destroyed. Since
+		// this is the latter lets not let etcd start and report the condition as an error.
+		// Result: error
+		if found && !dataDirExists {
+			return fmt.Errorf("member %q dataDir has been destoyed and should be removed from the cluster", target.String())
+		}
 
-	case err != nil:
-		fmt.Fprintf(os.Stderr, "Couldn't get targetMember. Returning error.\n")
-		return err
+		// Condition: target not found in member list but dataDir exists
+		// The member has been removed and should not start and return error. The dataDir will be archived once the operator
+		// scales up etcd.
+		// Result: error
+		if !found && dataDirExists {
+			return fmt.Errorf("member %q is no longer a member of the cluster and should not start", target.String())
+		}
 
-	case targetMember == nil && !memberDirExists:
-		// our member has not been added to the cluster and we have no previous data to start based on.
-		return fmt.Errorf("timed out")
+		// Condition: member not found
+		// The member list does not reflect the target member as it is waiting to be scaled up.
+		// Result: retry
+		if !found && !dataDirExists {
+			fmt.Fprintf(os.Stderr, "      member %q not found in member list, check operator logs for possible scaling problems\n", target.String())
```

Comment:
Should we return invalid initial cluster string to make it crash loop, so that it shows up in the pod status? 

Either here or at the end of the loop on L220.